src/Symfony/Component/HttpClient/NativeHttpClient.php

```diff
@@ -330,22 +330,36 @@ private static function parseHostPort(array $url, array &$info): array
      */
     private static function dnsResolve(string $host, NativeClientState $multi, array &$info, ?\Closure $onProgress): string
     {
-        $flag = '' !== $host && '[' === $host[0] && ']' === $host[-1] && str_contains($host, ':') ? \FILTER_FLAG_IPV6 : \FILTER_FLAG_IPV4;
-        $ip = \FILTER_FLAG_IPV6 === $flag ? substr($host, 1, -1) : $host;
+        if (filter_var(trim($host, '[]'), \FILTER_VALIDATE_IP)) {
+            return $host;
+        }
 
-        if (filter_var($ip, \FILTER_VALIDATE_IP, $flag)) {
-            // The host is already an IP address
-        } elseif (null === $ip = $multi->dnsCache[$host] ?? null) {
+        if (null === $ip = $multi->dnsCache[$host] ?? null) {
             $info['debug'] .= "* Hostname was NOT found in DNS cache\n";
             $now = microtime(true);
 
-            if (!$ip = gethostbynamel($host)) {
+            $tryResolveWithIpv6 = \defined('STREAM_PF_INET6');
```

Comment:
what about a wrapping `if (\defined('STREAM_PF_INET6')) {` instead of checking the variable many times?